### PR TITLE
Add Font Awesome 6 extractor

### PR DESCRIPTION
**Pull request by owenthewizard**
Based off of `nerdextractor.py`.

Open Questions:
* Should we fetch from something like [jsDelivr](https://www.jsdelivr.com/?docs=gh) instead of raw.githubusercontent.com? GitHub has rate limits/
* Do we want `fontawesome6extractor.py` to be versioned to Font Awesome 6 or generic?
* Naming: could shorten to `fa6extractor.py`

### CODE REVIEW

**Comment by owenthewizard** [comment]
Another thought, maybe we want to split this out into `fa6freeextractor`, `fa6proextractor`, `fa6brandsextractor`, etc?

**Comment by fdw** [comment]
Hey, this looks really good! Thank you 🙂

To answer your questions:

> Should we fetch from something like [jsDelivr](https://www.jsdelivr.com/?docs=gh) instead of raw.githubusercontent.com?

Since the extractors should not be called by most users at all and never automatically, I don't see rate limits as a problem. Also, I'd rather go right to the original source.

> Do we want fontawesome6extractor.py to be versioned to Font Awesome 6 or generic?

Versioning is fine, but as long as we only support one version, I don't see it as necessary.

> Naming: could shorten to fa6extractor.py

I'd rather have the full name, because I might not remember what "fa" is 😉

> Another thought, maybe we want to split this out into fa6freeextractor, fa6proextractor, fa6brandsextractor, etc?

Splitting the csv files is a good idea! Let's not show icons to people when the can't use them.
But I would only split the csvs, if possible, and keep the extractor as one.

Finally, can you add a call to the extractor to `main.py`?

**Comment by owenthewizard** [comment]
> Since the extractors should not be called by most users at all and never automatically, I don't see rate limits as a problem. Also, I'd rather go right to the original source.

Sounds good to me!

> Versioning is fine, but as long as we only support one version, I don't see it as necessary.

What happens when Fort Awesome releases Font Awesome 7? Also, if you or others are interested I could add FA 4 and 5 as well.

> Splitting the csv files is a good idea! Let's not show icons to people when the can't use them. But I would only split the csvs, if possible, and keep the extractor as one.

Great idea I'll do this.

> Finally, can you please add a call to the extractor to main.py?

👍

**Comment by fdw** [comment]
> > Versioning is fine, but as long as we only support one version, I don't see it as necessary.
> 
> What happens when Fort Awesome releases Font Awesome 7? Also, if you or others are interested I could add FA 4 and 5 as well.

I was going to answer "We'll find out if someone writes a new extractor", but then you already suggested doing that 😉 If you think old versions are still in use, let's have versioned csv files and possibly extractors 🙂

**Comment by fdw** [comment]
I've started to modernize and restructure the whole `extractors` module, which would lead to some (probably smaller) conflicts with your PR. Should I wait with pushing until we've merged this?

**Comment by owenthewizard** [comment]
> I've started to modernize and restructure the whole `extractors` module, which would lead to some (probably smaller) conflicts with your PR. Should I wait with pushing until we've merged this?

No go ahead, I'll rewrite as needed.

**Comment by fdw** [comment]
Done. Feel free to ask any questions you might have 🙂

**Comment by owenthewizard** [comment]
So it looks like only Font Awesome 6 Free + Brands are in the metadata so no need to split up CSVs.

P.S. with your recent commits, gitmoji extractor looks like it missed the `extract() -> extract_to()` change.

Could you write some documentation on the expectations for extractors, their expected usage, etc?

File: src/extractors/fontawesome6extractor.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            'https://raw.githubusercontent.com/FortAwesome/Font-Awesome/6.x/metadata/icons.json',
+            timeout=60
+        )
+        for (name, data) in response.json().items():
+            icon = int(data['unicode'], 16)
+            name = f'fa-{name}'
+            self.__icons.append(Character(icon, name))
+            try:
+                for alias in data['aliases']['names']:
+                    self.__icons.append(Character(icon, f'fa-{alias}'))
```
[the remark this answers is not in the document]
Comment (owenthewizard) [reply src/extractors/fontawesome6extractor.py:28]: Resolved in b4b82891008f45575a3315bacc9ecf96c20bfe7b.

File: src/extractors/__main__.py [hunk 2]
```diff
@@ -22,7 +23,8 @@
         CjkExtractor(),
         MathExtractor(character_factory),
         NerdFontExtractor(),
-        GitmojiExtractor()
+        GitmojiExtractor(),
+        FontAwesome6Extractor
```
[the remark this answers is not in the document]
Comment (owenthewizard) [reply src/extractors/__main__.py]: Resolved in b4b82891008f45575a3315bacc9ecf96c20bfe7b.

File: src/extractors/fontawesome6extractor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (owenthewizard) [reply src/extractors/fontawesome6extractor.py:28]: What would be the proper form for this, like in emojis?
```
 fa-face-smile <small>(fa-smile)</small>
```

Aside, it seems it's not entirely correct to call these CSV files. But that's a discussion for another day...

**Comment by owenthewizard** [comment]
> Looks good, thank you!
> 
> Can I also ask you to rebase onto `main` to remove/squash some commits?

Yes I'll go ahead and do that once the changes are reviewed and resolved.

File: src/extractors/fontawesome6extractor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fdw) [reply src/extractors/fontawesome6extractor.py:28]: Both this or just joining with `,` would be fine, I think. I haven't really used Font Awesomse, so I'm not sure if the aliases are fully equivalent to the name or not.

> Aside, it seems it's not entirely correct to call these CSV files. But that's a discussion for another day...

True. but `tsv` for tab-separated files never took off (and in the beginning, it was just tsv), and I didn't want to invent another extension.

[the remark this answers is not in the document]
Comment (owenthewizard) [reply src/extractors/fontawesome6extractor.py:28]: > Both this or just joining with , would be fine, I think.

Like this?
```csv
 fa-face-smile, fa-smile
 fa-address-book, fa-contact-book
 fa-address-card, fa-contact-card, fa-vcard
```

[the remark this answers is not in the document]
Comment (fdw) [reply src/extractors/fontawesome6extractor.py:28]: Works for me 🙂
But if you think ` fa-face-smile <small>(fa-smile)</small>` is better, that is also fine 🙂

[the remark this answers is not in the document]
Comment (owenthewizard) [reply src/extractors/fontawesome6extractor.py:28]: I think I prefer the way `<small> (alias) </small>` looks in `rofi` so I'll try that.

[the remark this answers is not in the document]
Comment (owenthewizard) [reply src/extractors/fontawesome6extractor.py]: How do you like it now? Not sure if that's the best way to work them in or not but it works.

File: src/extractors/fontawesome6extractor.py [hunk 3]
```diff
@@ -22,20 +22,28 @@ def __fetch_icons(self: 'FontAwesome6Extractor'):
...
     def __write_to_file(self, target: Path):
         if len(self.__icons) == 0:
             return
 
         with (target / 'fontawesome6.csv').open('w') as symbol_file:
             for icon in self.__icons:
-                symbol_file.write(f'{icon.directional_char} {html.escape(icon.name.lower())}\n')
+                line=f'{icon.directional_char} {html.escape(icon.name.lower())}'
+                if len(icon.aliases) > 0:
+                    line += ' <small>(' + ', '.join(icon.aliases) + ')</small>\n'
```
Comment (fdw) [inline src/extractors/fontawesome6extractor.py:42]: You could probably make it easier with `f" <small>{', '.join(icon.aliases)}</small>\n"`, if you want

**Review by fdw — APPROVED** [review]
Looks good, thank you!

**Comment by owenthewizard** [comment]
Squashed and ready for merge!

**Comment by fdw** [comment]
Awesome, thank you so much! 🙂

**Pull request merged**